src/util/amount.rs

```diff
@@ -0,0 +1,858 @@
+///
+/// This type implements several arithmetic operations from [std::ops].
+/// To prevent errors due to overflow or underflow when using these operations,
+/// it is advised to instead use the checked arithmetic methods whose names
+/// start with `checked_`.  The operations from [std::ops] that [Amount]
+/// implements will panic when overflow or underflow occurs.  Also note that
+/// since the internal representation of amounts is unsigned, subtracting below
+/// zero is considered an underflow and will cause a panic if you're not using
+/// the checked arithmetic methods.
+///
+#[derive(Copy, Clone, Hash)]
+pub struct Amount(Inner);
+// The Inner amount represents the number of satoshis.
+
+impl Amount {
+    /// The zero amount.
+    pub const ZERO: Amount = Amount(0);
+    /// Exactly one satoshi.
+    pub const ONE_SAT: Amount = Amount(1);
+    /// Exactly one bitcoin.
+    pub const ONE_BTC: Amount = Amount(100_000_000);
+
+    /// Create a new [Amount] using [amount] as the Inner type.
+    fn from_inner(amount: Inner) -> Amount {
+        Amount(amount)
+    }
+
+    /// Create an [Amount] with satoshi precision and the given number of satoshis.
+    pub fn from_sat(satoshi: u64) -> Amount {
+        Amount::from_inner(satoshi)
+    }
+
+    /// Get the number of satoshis in this [Amount].
+    pub fn as_sat(self) -> u64 {
+        self.0
+    }
+
+    /// The maximum value of an [Amount].
+    pub fn max_value() -> Amount {
+        Amount::from_inner(Inner::max_value())
+    }
+
+    // Don't use the Inner type in the methods below.
+    // Always use [Amount::from_sat] and [Amount::as_sat] instead.
+
+    /// Convert from a value expressing bitcoins to an [Amount].
+    pub fn from_btc(btc: f64) -> Result<Amount, ParseAmountError> {
+        Amount::from_float_in(btc, Denomination::Bitcoin)
+    }
+
+    /// Parse a decimal string as a value in the given denomination.
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value
+    /// with denomination, use [FromStr].
+    pub fn from_str_in(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
+        if s.len() == 0 {
+            return Err(ParseAmountError::InvalidFormat);
+        }
+        if s.len() > 50 {
+            return Err(ParseAmountError::InputTooLarge);
+        }
+
+        let max_decimals = {
+            // The difference in precision between native (satoshi)
+            // and desired denomination.
+            let precision_diff = -denom.precision();
+            if precision_diff < 0 {
+                // If precision diff is negative, this means we are parsing into a less
+                // precise amount.  That is not allowed unless there are no decimals and
+                // the last digits are zeroes as many as the diffence in precision.
+                let last_n = precision_diff.abs() as usize;
+                if s.contains(".") || !s.chars().skip(s.len() - last_n).all(|d| d == '0') {
```

Comment:
That might be personal preference, but to me
```rust
s.chars().rev().take(last_n).any(|d| d != '0')
```
looks much more straight forward than
```rust
!s.chars().skip(s.len() - last_n).all(|d| d == '0')
```

Reply:
Hmm, yeah I guess it's personal. I think the most important thing is that the comment above it explains what it is supposed to do. I think `all [...] == '0'` is more clear than `any [...] != '0'`. So how about a combination of both:
```rust
!s.chars().ref().take(last_n).all(|d| d == '0')
```

Reply:
meh, while typing I settled for you suggestion :D